### PR TITLE
Rewrite in TypeScript

### PR DESCRIPTION
**Pull request by orcachillin**
it was mentioned in a PR (#1) that you wanted to rewrite it in typescript, and I needed mii-js for a ts project, so I just went ahead and did it, happy to make any changes if wanted

### CODE REVIEW

File: .npmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+@NAMESPACE:registry=https://npm.pkg.github.com
```
Comment (jonbarrow) [inline .npmrc]: Why are you wanting to change the registry? We upload all of our packages to npm https://npmjs.com/org/pretendonetwork

File: .npmignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+src
+node_modules
+dist/tsbuildinfo
```
Comment (jonbarrow) [inline .npmignore]: npm will already use the `.gitignore` for ignoring files to publish, why is this here?

File: package.json [hunk 3]
```diff
@@ -1,18 +1,26 @@
 {
-  "name": "mii-js",
+  "name": "@gurrrrrrett3/miijs",
```
Comment (jonbarrow) [inline package.json]: We upload all of our packages to npm under the `@pretendonetwork` scope

File: package.json [hunk 4]
```diff
@@ -1,18 +1,26 @@
...
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "ci": "eslint . && tsc",
+    "build": "tsc"
   },
   "keywords": [],
   "author": "",
   "license": "GNU AFFERO GPLV3",
+  "repository": {
+    "type": "git",
+    "url": "git+github.com/gurrrrrrett3/mii-js.git"
```
Comment (jonbarrow) [inline package.json]: This is not correct

File: src/example.ts [hunk 5]
```diff
@@ -1,4 +1,4 @@
-const Mii = require('./mii');
+import Mii from "./mii";
```
Comment (jonbarrow) [inline src/example.ts]: This would not pass our linter. We use single quotes not double quotes

File: src/extendedBitStream.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    constructor(buffer: Buffer) {
+        super(buffer, buffer.byteOffset, buffer.byteLength);
+    }
+
+    public swapEndian() {
+        this.bigEndian = !this.bigEndian;
+    }
+
+    // the type definition for BitStream does not include the _index property
+    // since it's supposed to be private, but it's needed 4 times here sooo
```
Comment (jonbarrow) [inline src/extendedBitStream.ts]: The type definition can be augmented to include this, or you can probably try to just use `index`?

File: src/extendedBitStream.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
+import { BitStream } from "bit-buffer";
```
Comment (jonbarrow) [inline src/extendedBitStream.ts]: This would not pass our linter. We use single quotes not double quotes

File: src/extendedBitStream.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+    public writeUTF16String(string: string) {
+        const stringBuffer = Buffer.from(string, 'utf16le');
+        const terminatedBuffer = Buffer.alloc(0x14);
+
+        stringBuffer.copy(terminatedBuffer);
+
+        this.writeBuffer(terminatedBuffer);
+    }
+}
```
Comment (jonbarrow) [inline src/extendedBitStream.ts]: This would not pass our linter. We use tabs not spaces

**Review by jonbarrow — CHANGES_REQUESTED** [review]
There's linter and consistency issues with this. Additionally some changes were made with no real explanation, such as the ones related to NPM?

Also, was this tested? I've already started a TypeScript port of this library, and when trying to use `bit-buffer` it would fail to compile even without TypeScript errors in the IDE. Due to that, I've just started writing our own class for this.

Additionally, for consistency reasons this should really be set up the same way all of our other libraries are in terms of

- Linting
- Styles
- TypeScript configuration
- Build steps

For example https://github.com/PretendoNetwork/pictobox

File: .npmrc (hunk 1 above)
Comment (orcachillin) [reply .npmrc]: that was an accident, was meant to go to my fork

**Comment by orcachillin** [comment]
A bunch of these changes were pushed to the incorrect branch by me, will be reverting

**Comment by orcachillin** [comment]
I have tested this, see output: https://bin.gart.sh/60g9

**Comment by jonbarrow** [comment]
Looks fine to me now. There's a couple things I'd like to change but they're too minor for this PR, I'll just do them at some point later

**Review by jonbarrow — APPROVED** [review]
